### PR TITLE
Improve the performance of pull-request merging

### PR DESCRIPTION
**Pull request by typeless**
Merging pull-request without checking out a working tree

Inspired by [this](http://stackoverflow.com/a/10706075/3848102).

For #601

### CODE REVIEW

**Comment by appleboy** [comment]
Lint error, please fix lint error from the following command.

```sh
make lint
```

**Comment by typeless** [comment]
@appleboy I have fixed it in the last commit and tested with `make lint` for my local repo.

**Comment by appleboy** [comment]
@typeless Where is your last commit? Still fail: https://drone.gitea.io/go-gitea/gitea/1072

**Comment by typeless** [comment]
@appleboy https://github.com/go-gitea/gitea/pull/641/commits/0e58fa9cc6d71070f231d5e55de3df54bc578ad8

Am I missing something?

**Comment by appleboy** [comment]
@typeless I don't know why the drone show error on this commit. Please recommit to test it again.

**Comment by typeless** [comment]
@appleboy Alright

**Comment by Bwko** [comment]
@typeless Do you have any benchmarks? I'm using the latest master(without this PR) and merging a repo of 1.7gig succeeds in about 5 secs :hushed:

**Comment by typeless** [comment]
@Bwko For the record, the large repo that troubled me is about 13G the size of .git.

```
$ time git clone ../git-repositories/foo/my-large-repo.git/
Cloning into 'my-large-repo'...
done.
Checking out files: 100% (530225/530225), done.

real    5m58.914s
user    1m23.368s
sys     0m31.164s
```
The large repo is in fact an Android board support package which is a mixture of the Google official code and a vendor-tailored distribution.

And for Linux kernel, which is about 2G (bare repo)
```
$ time git clone gitea-repositories/foo/linux.git
Cloning into 'linux'...
done.
Checking out files: 100% (56233/56233), done.

real    0m6.242s
user    0m4.680s
sys     0m1.216s
```
Very close to what you have benchmarked.

The both clones were done within the same filesystem, by the way.

## Edit: 
For downloading the official android platform code:
```
$ curl https://storage.googleapis.com/git-repo-downloads/repo > ~/bin/repo
$ chmod a+x ~/bin/repo
```
Then use the `repo` tool to download the code
```
$ mkdir large-repo
$ cd large-repo
$ repo init -u https://android.googlesource.com/platform/manifest -b android-4.0.1_r1
$ repo sync -j32 
```
The `-j32` is similar to make -j, assuming your machine has 32 cores.
Be aware that this is a bandwidth killer. :trollface:

It'd better to kill the .repo and the sparsely populated .git directories before pushing to Gitea.
```
$ rm -rf .repo
$ find -name .git -type d -exec rm -rf {} +
```

References: https://source.android.com/source/downloading.html

**Comment by Bwko** [comment]
@typeless That looks really promising :+1: 
Found a bug, on your forked repo click new file, fill that file with contents use the `Create a new branch for this commit and start a pull request.` option.
An then create a PR, merging gives the following error:
```
[...routers/repo/pull.go:420 MergePullRequest()] [E] Merge: git commit-tree a9405d5ce9265ff2b5647b1f77ade6f44b6c419e -p master -p refs/pull/5/head: 
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.
```

**Comment by typeless** [comment]
Hasn't been tested yet :trollface:

**Comment by Bwko** [comment]
@typeless If you repeat the steps from my previous comment you'll get: ` fatal: invalid date format: RFC9999`

**Comment by typeless** [comment]
@Bwko Thanks for your help. 
I don't know if there is a better way but I would like to add an automated testing facility like how #741 does. I don't like to re-create those fake repos and PRs from the ground up again and again 😆

**Comment by lunny** [comment]
conflicted

**Comment by lunny** [comment]
@typeless any update?

**Comment by typeless** [comment]
@Bwko @metalmatze 
Rebased and manually resolved conflicts. PTAL.

P.S. The relavent commits: bf647ce14384e2b115843c6999437b02fe2da0e5 and d1006150fb3a82a8dd6e418578dc44474191bfd0

**Comment by lunny** [comment]
It's ready for review now?

**Comment by typeless** [comment]
@lunny I'd like to wait for #741 getting merged so that I can add an end-to-end test for this. For now, I just want to ping the related authors about what the PR does to their commits.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
So let's move this to v1.2 since we have no more time to review this one after #741 merged.

**Review by lunny — APPROVED** [review]

**Comment by typeless** [comment]
Given that @ethantkoenig is willing to take on the new integration test framework, I would like to let this PR open for reviewing. But, as mentioned before, I didn't test it much. 😆

**Comment by lunny** [comment]
Add some tests via the test framework?

**Comment by typeless** [comment]
@lunny When the new integration framework is mature enough, I will add tests based on it.

**Comment by typeless** [comment]
Tested it manually, found a bug and fixed it. I think this can be open for review.

**Comment by strk** [comment]
As #741 is now merged, can you please add tests for this ?

**Comment by typeless** [comment]
@strk #741 is going to be replaced by #1290. The complexity and effort of adding the test would depend on the infrastructure available. So I decided to wait and see (and reluctantly resort to manual testing). 

I have applied the PR to our own production instance for limited members with affordable risk. After all, the pull model is essentially broken for our large repos if this issue doesn't get fixed. And this has been complained more than once (including myself 😆 )

**Comment by lunny** [comment]
@typeless #1290 and #1627 should build the integration environment. Maybe you can continue this PR?

**Comment by typeless** [comment]
I'll add tests later.

Edit: found a problem, I'll look into it.

**Comment by lunny** [comment]
any news? or maybe we could move this to v1.3

**Comment by typeless** [comment]
@lunny Moving to v1.3 please. I need to spare some time for this.

**Comment by typeless** [comment]
Fixed the known bug.

**Comment by lunny** [comment]
@typeless maybe you can add some tests

**Comment by typeless** [comment]
I am still working on it.

**Comment by typeless** [comment]
The PR passes the test on my local machine, so it's ready for review now (if it still passes when the CI is fixed).

I would probably refactor the code into the `git` module in the future.

**Comment by lafriks** [comment]
Could you add some general comments like it was in old code? so it would be easier later to understand what exactly is going on in each part of the code

**Comment by typeless** [comment]
@lafriks Sure. The code has been updated.

**Comment by appleboy** [comment]
@typeless Please rebase the latest master branch or drone always build fail.

**Comment by typeless** [comment]
@appleboy I did it, unless I have mistaken something.

**Comment by appleboy** [comment]
@typeless My mistake. wait some moment.

**Comment by appleboy** [comment]
@typeless Please update again. Thanks.

**Comment by typeless** [comment]
@appleboy Done.

**Comment by appleboy** [comment]
@typeless Great work. build success. https://drone.gitea.io/go-gitea/gitea/3579

**Comment by lunny** [comment]
Any performance tests?

**Comment by typeless** [comment]
@lunny I was considering about it. But for showing any significance we need to prepare a large repo in the integration framework. I have no solid idea yet about how it should be added.

**Comment by lunny** [comment]
maybe clone linux repo on the tests?

**Comment by typeless** [comment]
@lunny maybe we can use the migration feature to clone a Linux repository for testing. But I am afraid of getting banned if we suck the bandwidth of the server too heavily/frequently. 😄

**Comment by lafriks** [comment]
I don't think performance tests should be done in drone

**Comment by sapk** [comment]
We could still have a benchmark suite at least to validate any performance PR. Those should not be run on drone even more if they are network/cpu/memory intensive.

**Comment by beddari** [comment]
Any status updates?

**Comment by lunny** [comment]
@typeless Is it possible to add a performance test?

**Comment by typeless** [comment]
I was thinking about it but not sure about how it should be done in an *idiomatic* way. Also the merge operation is stateful and not easily repeatable in a typical `for i:=0; i<b.N; i++ {}` loop. Any ideas? Maybe I am missing something.

**Comment by lunny** [comment]
maybe you can see https://github.com/go-gitea/gitea/pull/2519/files#diff-dbb0d04b955af59a55a562dc9e8c9912R39

File: models/pull.go
Comment (bkcsoft) [inline models/pull.go]: Does this hide the Email if the commiter has that set in their config?

Comment (typeless) [reply models/pull.go]: Yes. https://sourcegraph.com/github.com/go-gitea/gitea@6c6533fb31bbc82e7907b1dc233212045849569b/-/blob/models/user.go#L170

**Comment by typeless** [comment]
A benchmark has been added.

File: models/pull.go
Comment (lafriks) [inline models/pull.go:282]: Is not there missing `+pr.BaseRepo.Name+"-"+strconv.Itoa(time.Now().Nanosecond())`?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/pull.go
Comment (typeless) [reply models/pull.go:282]: ~~ioutil.TempDir will generate a suffix automatically. See https://golang.org/src/io/ioutil/tempfile.go?s=2138:2195#L66~~
Sorry, I see what you mean now. I'll fix it later.

Comment (typeless) [reply models/pull.go]: Ah, do you want a different prefix like `pr.BaseRepo.Name+`?

Comment (typeless) [reply models/pull.go:282]: >Is not there missing +pr.BaseRepo.Name+"-"+strconv.Itoa(time.Now().Nanosecond())?

This is actually correct despite a bit of inconsistent. I will clean this up a little bit.

The path `indexTmpPath` and `workTreeTmpPath` conrespond to `GIT_INDEX_FILE` and `GIT_WORK_TREE` respectively. I looked at this in a hurry last friday and got confused as well. 😆

**Comment by typeless** [comment]
This is actually not a comprehensive approach that covers all cases where `git-merge` can handle. So I think the change doesn't warrant the risks. The change should be made into the `git` tool or delegated to `libgit2' or src-d/go-git.

**Pull request closed without merging**
